### PR TITLE
Harden the v1 B-tree and local heap cache clients

### PR DESCRIPTION
**Pull request by derobins**
* Add bounds checks to the deserialize code
* Convert asserts to regular error checks
* General cleanup

### CODE REVIEW

File: src/H5Bcache.c [hunk 1]
```diff
@@ -156,57 +146,70 @@ H5B__cache_deserialize(const void *_image, size_t H5_ATTR_UNUSED len, void *_uda
...
+    /* Magic number */
+    if (H5_IS_BUFFER_OVERFLOW(image, H5_SIZEOF_MAGIC, p_end))
+        HGOTO_ERROR(H5E_BTREE, H5E_OVERFLOW, NULL, "ran off end of input buffer while decoding");
     if (HDmemcmp(image, H5B_MAGIC, (size_t)H5_SIZEOF_MAGIC) != 0)
         HGOTO_ERROR(H5E_BTREE, H5E_BADVALUE, NULL, "wrong B-tree signature")
     image += H5_SIZEOF_MAGIC;
 
-    /* node type and level */
+    /* Node type and level */
+    if (H5_IS_BUFFER_OVERFLOW(image, 1, p_end))
```
Comment (jhendersonHDF) [inline src/H5Bcache.c]: 2 instead of 1 here

File: src/H5Bcache.c [hunk 2]
```diff
@@ -156,57 +146,70 @@ H5B__cache_deserialize(const void *_image, size_t H5_ATTR_UNUSED len, void *_uda
...
+    /* Child/key pairs */
     native = bt->native;
     for (u = 0; u < bt->nchildren; u++) {
         /* Decode native key value */
+        if (H5_IS_BUFFER_OVERFLOW(image, shared->sizeof_rkey, p_end))
+            HGOTO_ERROR(H5E_BTREE, H5E_OVERFLOW, NULL, "ran off end of input buffer while decoding");
         if ((udata->type->decode)(shared, image, native) < 0)
             HGOTO_ERROR(H5E_BTREE, H5E_CANTDECODE, NULL, "unable to decode key")
         image += shared->sizeof_rkey;
         native += udata->type->sizeof_nkey;
```
Comment (jhendersonHDF) [inline src/H5Bcache.c:199]: Should we be tracking whether we run off the end of the native buffer as well, or is that handled elsewhere?

File: src/H5Bcache.c (hunk 1 above)
Comment (derobins) [reply src/H5Bcache.c]: Fixed

File: src/H5Bcache.c (hunk 2 above)
Comment (derobins) [reply src/H5Bcache.c:199]: For this pass, I was just worried about the image since that's where most of the CVE issues come from.

**Review by jhendersonHDF — APPROVED** [review]

**Review by byrnHDF — APPROVED** [review]

**Pull request merged**